go/test/endtoend/vtgate/queries/dml/insert_test.go

```diff
@@ -396,6 +396,74 @@ func TestInsertSelectUnshardedUsingSharded(t *testing.T) {
 	}
 }
 
+// MySQL returns 0 as the `LastInsertId` on `... ON DUPLICATE KEY UPDATE ...` queries
+// when no new row is inserted. This test verifies that Vitess behaves the same way.
+func TestInsertShardedWithOnDuplicateKeyNoInserts(t *testing.T) {
+	mcmp, closer := start(t)
+	defer closer()
+
+	mcmp.Exec("insert into last_insert_id_test (id, sharding_key, user_id, reason) values (1, '1:1:1', 1, 'foo'), (2, '2:2:2', 2, 'bar')")
+
+	// Bump the sequence value so the sequence accounts for the 2 explicit inserts above.
+	utils.Exec(t, mcmp.VtConn, "SELECT NEXT 2 VALUES FROM uks.last_insert_id_test_seq")
+
+	// First test case, insert a row that already exists, and don't actually change any column at all.
+	query := "insert into last_insert_id_test (sharding_key, user_id, reason) values ('1:1:1', 1, 'foo') on duplicate key update reason = reason"
+
+	mysqlResult := utils.Exec(t, mcmp.MySQLConn, query)
+	// no new row inserted, so insert id should be 0.
+	assert.Equal(t, uint64(0), mysqlResult.InsertID)
+	// no row was modified, so rows affected should be 0.
+	assert.Equal(t, uint64(0), mysqlResult.RowsAffected)
+
+	vitessResult := utils.Exec(t, mcmp.VtConn, query)
+	assert.Equal(t, mysqlResult.RowsAffected, vitessResult.RowsAffected)
+	assert.Equal(t, mysqlResult.InsertID, vitessResult.InsertID)
+
+	// Second test case, insert a row that already exists, and change a column on the existing row.
+	query = "insert into last_insert_id_test (sharding_key, user_id, reason) values ('1:1:1', 1, 'bar') on duplicate key update reason = VALUES(reason)"
+
+	mysqlResult = utils.Exec(t, mcmp.MySQLConn, query)
+	// a row was modified, so insert id should match the auto increment column value of the modified row
+	assert.Equal(t, uint64(1), mysqlResult.InsertID)
+	// one row was modified, so rows affected should be 2.
+	assert.Equal(t, uint64(2), mysqlResult.RowsAffected)
+
+	vitessResult = utils.Exec(t, mcmp.VtConn, query)
+	assert.Equal(t, mysqlResult.RowsAffected, vitessResult.RowsAffected)
+	// Vitess can't return the `auto_increment` value of the last updated row, but it returns a value larger than 0.
+	assert.Greater(t, vitessResult.InsertID, uint64(0))
```

Comment:
We do not need to test with way, mcmp.Exec can validate all of this. 
With latest changes to this test package we can add new compare option